# KNN

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np 
%matplotlib inline
import scipy as sp
# increase figure size
plt.rcParams['figure.figsize'] = [15, 10]

## KNN Grid Search Results

First, we'll run scale_run.py and grid_search.py in the command line. Then we'll import the raw results of the grid search to identify the relevant columns.

In [2]:
tree_results = pd.read_csv('../tmp/grid_search_knn0.csv').drop(columns=['Unnamed: 0'])
tree_results.head(10)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.020488      0.007085         0.023287        0.007290   
1       0.016989      0.003345         0.023487        0.010155   
2       0.015090      0.002299         0.013893        0.000700   
3       0.014791      0.001989         0.014347        0.001882   
4       0.017389      0.005621         0.019190        0.005397   
5       0.016490      0.002800         0.018690        0.001845   
6       0.017088      0.001969         0.021293        0.007491   
7       0.016789      0.002560         0.021788        0.005773   
8       0.017090      0.003046         0.019290        0.002721   
9       0.019588      0.007770         0.020190        0.002924   

  param_classifier__algorithm  param_classifier__leaf_size  \
0                        auto                           15   
1                        auto                           15   
2                        auto                           15   
3                        auto                           15   
4                        auto                           15   
5                        auto                           15   
6                        auto                           15   
7                        auto                           15   
8                        auto                           15   
9                        auto                           15   

  param_classifier__metric  param_classifier__n_neighbors  \
0                minkowski                              3   
1                minkowski                              3   
2                minkowski                              3   
3                minkowski                              3   
4                minkowski                              5   
5                minkowski                              5   
6                minkowski                              5   
7                minkowski                              5   
8                minkowski                              7   
9                minkowski                              7   

   param_classifier__p param_classifier__weights         ...          \
0                    1                   uniform         ...           
1                    1                  distance         ...           
2                    2                   uniform         ...           
3                    2                  distance         ...           
4                    1                   uniform         ...           
5                    1                  distance         ...           
6                    2                   uniform         ...           
7                    2                  distance         ...           
8                    1                   uniform         ...           
9                    1                  distance         ...           

  split2_train_accuracy  split3_train_accuracy  split4_train_accuracy  \
0              0.684644               0.701873               0.702096   
1              1.000000               1.000000               1.000000   
2              0.686142               0.700375               0.693114   
3              1.000000               1.000000               1.000000   
4              0.606742               0.621723               0.624251   
5              1.000000               1.000000               1.000000   
6              0.605993               0.628464               0.622006   
7              1.000000               1.000000               1.000000   
8              0.569288               0.579775               0.582335   
9              1.000000               1.000000               1.000000   

   split5_train_accuracy  split6_train_accuracy  split7_train_accuracy  \
0               0.700599               0.696856               0.693862   
1               1.000000               1.000000               1.000000   
2               0.697605               0.693114               0.695359   
3               1

## KNN Step 0 Scores

Now we will narrow down the relevant columns and sort by accuracy score.

In [3]:
columns = ['rank_test_accuracy', 
           'mean_test_accuracy', 'mean_train_accuracy',
           'std_test_accuracy', 'std_train_accuracy',
           'param_classifier__n_neighbors', 'param_classifier__weights', 'param_classifier__algorithm',
           'param_classifier__leaf_size', 'param_classifier__p',
           'param_classifier__metric']
tree_results.sort_values('rank_test_accuracy')[columns].head(30)

rank_test_accuracy  mean_test_accuracy  mean_train_accuracy  \
546                   1            0.474394             0.695118   
578                   1            0.474394             0.695118   
514                   1            0.474394             0.695118   
482                   1            0.474394             0.695118   
610                   1            0.474394             0.695118   
483                   6            0.473046             1.000000   
547                   6            0.473046             1.000000   
515                   6            0.473046             1.000000   
611                   6            0.473046             1.000000   
579                   6            0.473046             1.000000   
386                  11            0.471024             0.693845   
259                  11            0.471024             1.000000   
34                   11            0.471024             0.693770   
163                  11            0.471024             1.000000   
98                   11            0.471024             0.693845   
387                  11            0.471024             1.000000   
258                  11            0.471024             0.693995   
418                  11            0.471024             0.693845   
355                  11            0.471024             1.000000   
99                   11            0.471024             1.000000   
419                  11            0.471024             1.000000   
227                  11            0.471024             1.000000   
226                  11            0.471024             0.693995   
194                  11            0.471024             0.693695   
195                  11            0.471024             1.000000   
66                   11            0.471024             0.693845   
67                   11            0.471024             1.000000   
354                  11            0.471024             0.693770   
162                  11            0.471024             0.693695   
35                   11            0.471024             1.000000   

     std_test_accuracy  std_train_accuracy  param_classifier__n_neighbors  \
546           0.043879            0.005973                              3   
578           0.043879            0.005973                              3   
514           0.043879            0.005973                              3   
482           0.043879            0.005973                              3   
610           0.043879            0.005973                              3   
483           0.043794            0.000000                              3   
547           0.043794            0.000000                              3   
515           0.043794            0.000000                              3   
611           0.043794            0.000000                              3   
579           0.043794            0.000000                              3   
386           0.045648            0.005676                              3   
259           0.044738            0.000000                              3   
34            0.045648            0.005771                              3   
163           0.044738            0.000000                              3   
98            0.045648            0.005676                              3   
387           0.044738            0.000000                              3   
258           0.045648            0.006001                              3   
418           0.045648            0.005676                              3   
355           0.044738            0.000000                              3   
99            0.044738            0.000000                              3   
419           0.044738            0.000000                              3   
227           0.044738            0.000000                              3   
226           0.045648            0.006001                              3   
194           0.045648            0.005940      

Apparently, the leaf size doesn't have much of an effect on the score, so we will leave it at the default setting of 30.
The results suggest that 3 is the ideal value for the n_neighbors parameter and 2 is the ideal value for the p parameter.

The weights and algorithm parameters are still mixed, so we will fix the other parameters and try again.

## KNN Step 1 Scores

In [4]:
tree_results = pd.read_csv('../tmp/grid_search_knn1.csv')
columns = ['rank_test_accuracy', 
           'mean_test_accuracy', 'mean_train_accuracy',
           'std_test_accuracy', 'std_train_accuracy',
           'param_classifier__n_neighbors', 'param_classifier__weights', 'param_classifier__algorithm',
           'param_classifier__leaf_size', 'param_classifier__p',
           'param_classifier__metric']
tree_results.sort_values('rank_test_accuracy')[columns].head(10)

rank_test_accuracy  mean_test_accuracy  mean_train_accuracy  \
6                   1            0.474394             0.695118   
7                   2            0.473046             1.000000   
0                   3            0.471024             0.693845   
1                   3            0.471024             1.000000   
2                   3            0.471024             0.693995   
3                   3            0.471024             1.000000   
4                   3            0.471024             0.693845   
5                   3            0.471024             1.000000   

   std_test_accuracy  std_train_accuracy  param_classifier__n_neighbors  \
6           0.043879            0.005973                              3   
7           0.043794            0.000000                              3   
0           0.045648            0.005676                              3   
1           0.044738            0.000000                              3   
2           0.045648            0.006001                              3   
3           0.044738            0.000000                              3   
4           0.045648            0.005676                              3   
5           0.044738            0.000000                              3   

  param_classifier__weights param_classifier__algorithm  \
6                   uniform                       brute   
7                  distance                       brute   
0                   uniform                        auto   
1                  distance                        auto   
2                   uniform                   ball_tree   
3                  distance                   ball_tree   
4                   uniform                     kd_tree   
5                  distance                     kd_tree   

   param_classifier__leaf_size  param_classifier__p param_classifier__metric  
6                           30                    2                minkowski  
7                           30                    2                minkowski  
0                           30                    2                minkowski  
1                           30                    2                minkowski  
2                           30                    2                minkowski  
3                           30                    2                minkowski  
4                           30                    2                minkowski  
5                           30                    2                minkowski

It looks like 'brute' is the best choice for the algorithm and 'uniform' the best choice for weight at this point.

So the final parameters that we will be using for KNN are:  
n_neighbors = 3  
weights = ‘uniform’  
algorithm = ‘brute’  
leaf_size = 30  
p = 2  
metric = ‘minkowski’  

This combination yielded the follow scores:  
mean_train_accuracy = 0.695118  
mean_test_accuracy = 0.474394  
std_test_accuracy = 0.043879  
std_train_accuracy = 0.005973  

The lack of deviation in the test and train scores indicates that the scoring is consistent. The imbalance in the mean scores is concerning though. The mean score of the training data is at about 70%, while the mean score of the test data is at about 50%. This means that the model isn't generalizing enough and is overfitting to a certain degree. 